docs:  introduction sections and nits (#2200)
hatchet-dev/hatchet: frontend/docs/pages/home/guarantees-and-tradeoffs.mdx

import { Callout } from "nextra/components";

# Guarantees & Tradeoffs

Hatchet is designed as a modern task orchestration platform that bridges the gap between simple job queues and complex workflow engines. Understanding where it excels—and where it doesn't—will help you determine if it's the right fit for your needs.

### Good Fit

<table>
  <tbody>
    <tr>
      <td>✅</td>
      <td>

>         <strong>Real-time Requests</strong> - Sub-50ms task dispatch for hot

Reviewer: I think we have some conflicting information (or at least some possibly confusion-causing language) in different places - we quote 35ms, 100ms, and 50ms dispatch times / start times

        workers with thousands of concurrent tasks
      </td>
    </tr>
    <tr>
      <td>✅</td>
      <td>
        <strong>Workflow Orchestration</strong> with dependencies and error
        handling
      </td>
    </tr>
    <tr>
      <td>✅</td>
      <td>
        <strong>Reliable Task Processing</strong> where durability matters
      </td>
    </tr>
    <tr>
      <td>✅</td>
      <td>

>         <strong>Moderate Throughput</strong> (hundreds to low 10,000s of

Reviewer: in the intro, we mention billions of runs per month, so I think this is a bit confusing too (an outsider might read this and think "which is it?")

        tasks/second)
      </td>
    </tr>
    <tr>
      <td>✅</td>
      <td>

>         <strong>Multi-Language Workers</strong> or polyglot teams

Reviewer: ```suggestion
        <strong>Multi-Language Workers</strong> for polyglot teams
```

      </td>
    </tr>
    <tr>
      <td>✅</td>
      <td>
        <strong>Operational Simplicity</strong> if your team is already using
        PostgreSQL
      </td>
    </tr>
    <tr>
      <td>✅</td>
      <td>

>         <strong>Cloud or Air-gapped Environments</strong> for flexible

Reviewer: ```suggestion
        <strong>Cloud or Air-Gapped Environments</strong> for flexible
```

</table>

### Not a Good Fit

<table>
  <tbody>
    <tr>
      <td>❌</td>
      <td>
        <strong>Extremely High Throughput</strong> (consistently 10,000+
        tasks/second)
      </td>
    </tr>
    <tr>
      <td>❌</td>
      <td>
        <strong>Sub-Millisecond Latency</strong> requirements
      </td>
    </tr>
    <tr>
      <td>❌</td>
      <td>
        <strong>Simple Queuing</strong> without workflow features
      </td>
    </tr>
    <tr>
      <td>❌</td>
      <td>
        <strong>Memory-Only Queuing</strong> where persistence or durability
        isn't needed
      </td>
    </tr>
    <tr>
      <td>❌</td>
      <td>
        <strong>Serverless Environments</strong> on cloud providers like AWS
        Lambda, Google Cloud Functions, or Azure Functions
      </td>
    </tr>
  </tbody>
</table>

## Core Reliability Guarantees

Hatchet is designed with the following core reliability guarantees:

**Every task will execute at least once.** Hatchet ensures that no task gets lost, even during system failures, network outages, or deployments. Failed tasks automatically retry according to your configuration, and all tasks persist through restarts and network issues.

> **Consistent state management.** All workflow state changes happen within PostgreSQL transactions, ensuring that your workflow dependencies resolve consistently and no tasks are lost during failures or deployments. This ACID consistency means your workflows maintain integrity even under adverse conditions.

Reviewer: nit: ACID is a bit of an abused term

Author: yeah fair

**Predictable execution order.** The default task assignment strategy is First In First Out (FIFO) which can be modified with [concurrency policies](./concurrency.mdx), [rate limits](./rate-limits.mdx), and [priorities](./priority.mdx).

**Operational resilience.** The engine and API servers are stateless, allowing them to restart without losing state and enabling horizontal scaling by simply adding more instances. Workers automatically reconnect after network issues and can be deployed anywhere—containers, VMs, or local development environments.

## Performance Expectations

Understanding Hatchet's performance characteristics helps you plan your implementation and set realistic expectations.

**Typical time-to-start latency** for task dispatch is sub 100ms with PostgreSQL storage, though this can be optimized to ~35ms P95 for hot workers in optimized setups. Network latency between your workers and the Hatchet engine will directly impact dispatch times, so consider deployment topology when latency matters.

> **Throughput capacity** varies significantly based on your setup. A single engine instance with PostgreSQL-only storage typically handles hundreds of tasks per second. When you need higher throughput, adding RabbitMQ as a message queue can substantially increase capacity, though your database will eventually become the bottleneck at very high scales. Through tuning and sharding, we can support throughputs of 10,000s of tasks per second.

Reviewer: ```suggestion
**Throughput capacity** varies significantly based on your setup. A single engine instance with PostgreSQL-only storage typically handles hundreds of tasks per second. When you need higher throughput, adding RabbitMQ as a message queue can substantially increase capacity, though your database will eventually become the bottleneck at very high scales. Through tuning and sharding, we can support throughputs of tens of thousands of tasks per second.
```

> **Concurrent processing** scales well—Hatchet supports thousands of concurrent workers, with worker-level concurrency controlled through slot configuration. The depth of your queues is limited by your database storage capacity rather than memory constraints.

Reviewer: ```suggestion
**Concurrent processing** scales well — Hatchet supports thousands of concurrent workers, with worker-level concurrency controlled through slot configuration. The depth of your queues is limited by your database storage capacity rather than memory constraints.
```
